### ISSUE TITLE
Static analyzer cplusplus.NewDeleteLeaks false positive on placement new operator

### ISSUE THREAD

**Issue by ptomato** [llvm/llvm-project#195557]
I have a false positive where CSA reports "Potential memory leak" but no memory is allocated. This originally occurred using Firefox code, but I've created a standalone test case. Note that triggering the false positive is very sensitive to specific constructs in the code, so the test case is long and contains many lines that look like they should be unnecessary.

## Test case

**a.cpp**:
```c++
// clang++ --analyze -Xanalyzer -analyzer-output=text -isystem . a.cpp
// (-isystem is necessary, not -I)

#include <type_traits>
#include <utility>

#include <a.h>  // angle brackets are necessary

void frob();

struct Opaque;
void frobOpaque(const Opaque* o) {
    if (o)
        frob();
}

// Greatly reduced version of mozilla::Maybe from Firefox codebase
template <class T>
class Maybe {
    union Storage {
        std::remove_const_t<T> val;
        Storage() {}
        ~Storage() {}
    } m_storage;
    bool m_is_some = false;

    template <typename... Args>
    constexpr void emplace(Args&&... args) {
        ::new (NotNullTag{}, &m_storage.val) T(std::forward<Args>(args)...);
        m_is_some = true;
    }

 public:
    ~Maybe() {
        if (m_is_some)
            m_storage.val.~T();
    }

    Maybe(const Maybe& other) {
        if (other.is_some())
            emplace(*other);
    }

    template <typename U>
    Maybe(const Maybe<U>& other) {
        if (other.is_some())
            emplace(*other);
    }

    constexpr bool is_some() const { return m_is_some; }
    constexpr const T& operator*() const& { return m_storage.val; }
};

struct S1 {
    int i;
    void frob() const {
        if (i)
            frobOpaque(reinterpret_cast<Opaque*>(i));
    }
};

struct S2 {};

struct S3 {
    S3(const S3&) {}  // this cannot be =default
    operator S2() const { return {}; }
};

struct S4 {
    Maybe<S3> m_data;
    Maybe<S3> data() const { return m_data; }
};

void foo(const Maybe<S2>&, S1[3]);

void bar(S1 s1[3], S4* s4) {
    // All three calls to frob() are necessary, as well as passing s1 to foo()
    s1[0].frob();
    s1[1].frob();
    s1[2].frob();
    foo(s4->data(), s1);
}
```

**a.h**:
```c++
#pragma once

#include <stddef.h>
#include <stdlib.h>

inline void nothing() {}

struct NotNullTag {};

inline void* operator new(size_t, NotNullTag, void* p) {
    // Yes, all of these do-whiles are necessary, as well as the nothing() call
    // (mimics MOZ_ASSERT from Firefox codebase)
    do {
        if (!p) {
            do {
            } while (false);
            nothing();
            do {
            } while (false);
            do {
                abort();
            } while (false);
        }
    } while (false);
    return p;
}
```

Trigger with
```
clang++ --analyze -Xanalyzer -analyzer-output=text -isystem . a.cpp
``` 
or alternatively using clang-tidy
```
clang-tidy --checks='-*,clang-analyzer-cplusplus.NewDeleteLeaks' a.cpp -- -isystem .
```

## Expected output

None.

## Actual output

```
a.cpp:30:19: warning: Potential memory leak [cplusplus.NewDeleteLeaks]
   30 |         m_is_some = true;
      |                   ^
a.cpp:81:9: note: Calling constructor for 'Maybe<S2>'
   81 |     foo(s4->data(), s1);
      |         ^~~~~~~~~~
a.cpp:46:9: note: Taking true branch
   46 |         if (other.is_some())
      |         ^
a.cpp:47:13: note: Calling 'Maybe::emplace'
   47 |             emplace(*other);
      |             ^~~~~~~~~~~~~~~
a.cpp:29:9: note: Memory is allocated
   29 |         ::new (NotNullTag{}, &m_storage.val) T(std::forward<Args>(args)...);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
a.cpp:30:19: note: Potential memory leak
   30 |         m_is_some = true;
      |                   ^
1 warning generated.
```

This can't be a memory leak because no memory is being allocated. In addition, changing the surrounding code just slightly (e.g., removing one of the `do {} while (false)` in `a.h`) causes the false positive to no longer trigger.

## Version data

```
$ clang++ --version
clang version 22.1.4 (Fedora 22.1.4-1.fc44)
Target: x86_64-redhat-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
Configuration file: /etc/clang/x86_64-redhat-linux-gnu-clang++.cfg
$ cat /etc/clang/x86_64-redhat-linux-gnu-clang++.cfg
 --gcc-triple=x86_64-redhat-linux
```
